util/tippingContractUtil.js

```diff
@@ -2,25 +2,82 @@ const BigNumber = require('bignumber.js');
 
 const tippingContractUtil = {};
 
-tippingContractUtil.getTipsRetips = (state) => {
+tippingContractUtil.getTipsRetips = (...states) => {
+  if (!Array.isArray(states) || (Array.isArray(states) && states.length === 0)) throw Error("states must be passed as additional arguments");
+
+  // not very performant nesting of reduces
+  return states.reduce((acc, cur) => {
+    if (!cur.result || !cur.decodedResult) throw Error("full returned tx state must be passed");
+    const aggregation = aggregateState(cur);
+    acc.urls = aggregation.urls.reduce((accUrls, curUrl) => {
+      let returnUrls = accUrls;
+      const accHasCurUrl = accUrls.find(a => a.url === curUrl.url);
+
+      // if has url in accumulator, filter it out and replace with updated values, otherwise just add to accumulator
+      if (accHasCurUrl) {
+        returnUrls = returnUrls.filter(a => a.url !== curUrl.url);
+        returnUrls.push({
+          url: curUrl.url,
+          tip_ids: accHasCurUrl.tip_ids.concat(curUrl.tip_ids),
+          retip_ids: accHasCurUrl.retip_ids.concat(curUrl.retip_ids),
+          unclaimed_amount: new BigNumber(accHasCurUrl.unclaimed_amount).plus(curUrl.unclaimed_amount).toFixed(),
+          token_unclaimed_amount: accHasCurUrl.token_unclaimed_amount.concat(curUrl.token_unclaimed_amount)
+            .reduce((accToken, curToken) => {
+              const hasToken = accToken.find(a => a.token === curToken.token)
+              if (hasToken) {
+                accToken = accToken.filter(a => a.token !== curToken.token);
+                accToken.push({
+                  token: curToken.token,
+                  amount: new BigNumber(hasToken.amount).plus(curToken.amount).toFixed()
+                })
+              } else {
+                accToken.push(curToken);
+              }
+
+              return accToken;
+            }, [])
```

Comment:
we could make this a lot less costly (removing find & filter) by using an object

```suggestion
            .reduce((accToken, curToken) => {
              var oldAmount = accToken[curToken.token] ? accToken[curToken.token].amount : 0;
              accToken[curToken.token] = new BigNumber(oldAmount).plus(curToken.amount);
              return accToken;
            }, {})
            
            // TODO potentially convert to array?
```

Reply:
the returned datastructure from this is different, will be harder to work with this in frontend

Reply:
I had it like this previously but it didn't work out

Reply:
I think its still a lot faster, even if we convert it back to the original structure by iterating over the object once with `Object.keys().map`

Reply:
right, I even do this in other parts of the code already

Reply:
adjusted here: https://github.com/aeternity/tipping-contract/pull/17/commits/b0afa1b05c92dbbe672d5fcf15ff4cde4b5fcdda